engine/table/src/main/java/io/deephaven/engine/table/impl/AbstractFilterExecution.java

```diff
@@ -48,273 +43,287 @@ abstract class AbstractFilterExecution extends AbstractNotification {
                 }
+            } else {
+                adds = null;
             }
-            if (modifyInput != null) {
-                if (Thread.interrupted()) {
-                    throw new CancellationException("interrupted while filtering");
-                }
-                try (final RowSet processModifies = modifyInput.subSetByPositionRange(modifyStart, modifyEnd)) {
-                    modifyResult = filters[filterIndex].filter(
-                            processModifies, sourceTable.getRowSet(), sourceTable, usePrev);
+            if (modsToUse != null) {
+                try (final RowSet processMods = modsToUse.subSetByPositionRange(modifiedStart, modifiedEnd)) {
+                    mods = filter.filter(processMods, sourceTable.getRowSet(), sourceTable, usePrev);
                 }
+            } else {
+                mods = null;
             }
-            if (Thread.interrupted()) {
-                throw new CancellationException("interrupted while filtering");
-            }
-            scheduleNextFilter(onCompletion);
+            onComplete.accept(adds, mods);
         } catch (Exception e) {
-            exceptionResult = e;
-            onCompletion.accept(this);
+            onError.accept(e);
         }
     }
 
-    RowSet getAddedResult() {
-        return addedResult == null ? RowSetFactory.empty() : addedResult;
-    }
-
-    RowSet getModifyResult() {
-        return modifyResult == null ? RowSetFactory.empty() : modifyResult;
-    }
-
     /**
-     * Collapse other's result into this AbstractFilterExecution's result.
+     * Run the single filter specified by this AbstractFilterExecution and store the results in addedResult and
+     * modifyResult. Processes all rows in the added and modified inputs.
      *
-     * @param other the result to combine into this
+     * @param filter the filter to execute
+     * @param addsToUse the added input to use for this filter
+     * @param modsToUse the modified input to use for this filter
+     * @param onComplete the routine to call after the filter has been successfully executed
+     * @param onError the routine to call if a filter raises an exception
      */
-    void combine(AbstractFilterExecution other) {
-        if (this.addedResult == null) {
-            this.addedResult = other.addedResult;
-        } else if (other.addedResult != null) {
-            this.addedResult.insert(other.addedResult);
-        }
-        if (this.modifyResult == null) {
-            this.modifyResult = other.modifyResult;
-        } else if (other.modifyResult != null) {
-            this.modifyResult.insert(other.modifyResult);
+    private void doFilter(
+            final WhereFilter filter,
+            final RowSet addsToUse,
+            final RowSet modsToUse,
+            final FilterComplete onComplete,
+            final Consumer<Exception> onError) {
+        if (Thread.interrupted()) {
```

Comment:
I'm a little torn on making this suggestion. It feels like this should defer to the other method. The `subSetByPositionRange` will detect that the appropriate thing to do is to do a `copy()`. On the other hand, both methods are super straightforward and this implementation is ever-so-slightly faster.

Leaving the snippet just for consideration. Maybe Ryan will have a strong preference.
```java
        doFilter(filter, addsToUse, 0, addsToUse == null ? 0 : addsToUse.size(),
                modsToUse, 0, modsToUse == null ? 0 : modsToUse.size(),
                onComplete, onError);
```